File: android/app/src/main/kotlin/com/mytiki/app/ZendeskApi.kt
package com.mytiki.app

import android.content.Context
import com.zendesk.logger.Logger
import com.zendesk.service.ErrorResponse
import com.zendesk.service.ZendeskCallback
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject
import zendesk.core.AnonymousIdentity
import zendesk.core.Identity
import zendesk.core.Zendesk
import zendesk.support.*

typealias ListCallback<T> = (T) -> Unit
typealias SingleCallback<T> = (T) -> Unit
typealias ErrorCallback = (ErrorResponse) -> Unit

class ZendeskApi {

    private lateinit var providerStore: ProviderStore
    private lateinit var helpCenterProvider: HelpCenterProvider

    fun initZendesk(context : Context) {
        val url = "https://mytikihelp.zendesk.com/"
        val appId = "12abbc10c46d1731876d529d0b90656cc81d416a9452d0be"
        val clientId = "mobile_sdk_client_3beab601d34d787fc05c"

        Logger.setLoggable(true);
        Zendesk.INSTANCE.init(context,
                url,
                appId,
                clientId)
        Support.INSTANCE.init(Zendesk.INSTANCE)
        val identity: Identity = AnonymousIdentity.Builder().build()
        Zendesk.INSTANCE.setIdentity(identity)
        providerStore = Support.INSTANCE.provider()!!
        helpCenterProvider = providerStore.helpCenterProvider()
    }

    fun getZendeskCategories(result: MethodChannel.Result) {
        getCategories(
            {categories -> result.success(categories)},
            {error -> result.error(error.status.toString(), error.reason, error.toString())}
        )
    }

    fun getZendeskSections(call: MethodCall, result: MethodChannel.Result) {
        val categoryID = if ( call.hasArgument("categoryId") ) call.argument<Any?>("categoryId" )
        else result.error("400", "categoryId argument is missing", null)
        getSections(
            categoryID as Long,
            {sections -> result.success(sections)},
            {error -> result.error(error.status.toString(), error.reason, error)}
        )
    }

    fun getZendeskArticles(call: MethodCall, result: MethodChannel.Result) {
        val sectionId = if ( call.hasArgument("sectionId") ) call.argument<Any?>("sectionId" )
        else result.error("400", "sectionId argument is missing", null)
        getArticles(
            sectionId as Long,
            {sections -> result.success(sections)},
            {error -> result.error(error.status.toString(), error.reason, error)}
        )
    }

    fun getZendeskArticle(call: MethodCall, result: MethodChannel.Result) {
        val articleId = if ( call.hasArgument("articleId") ) call.argument<Any?>("articleId" )
        else result.error("400", "articleId argument is missing", null)
        getArticle(
            articleId as Long,
            {sections -> result.success(sections)},
            {error -> result.error(error.status.toString(), error.reason, error)}
        )
    }

    private fun getCategories(onSuccess: ListCallback<ArrayList<Map<String, Any>>>, onError: ErrorCallback){
        helpCenterProvider.getCategories( object: ZendeskCallback<List<Category>>(){
            override fun onSuccess(categories: List<Category>){
                val cats : ArrayList<Map<String,Any>> = ArrayList<Map<String, Any>>()
                categories.forEach{
                    cats.add(processCategory(it)!!)
                }
                onSuccess(cats)
            }
            override fun onError(error: ErrorResponse){
                onError(error)
            }
        })
    }

    private fun processCategory(it: Category): Map<String, Any>? {
        if(it.id != null && it.name != null && it.description != null){
            return mapOf(
                    "id" to it.id!!,
                    "title" to it.name!!,
                    "description" to it.description!!
            )
        }
        return null
    }

    private fun getSections(category: Long, onSuccess: ListCallback<ArrayList<Map<String, Any>?>>, onError: ErrorCallback){
        helpCenterProvider.getSections( category, object: ZendeskCallback<List<Section>>(){
            override fun onSuccess(sections: List<Section>){
                val sectionList : ArrayList<Map<String, Any>?> = ArrayList<Map<String, Any>?>()
                sections.forEach{
                    sectionList.add(
                        processSection(it)
                    )
                }
                onSuccess(sectionList)
            }
            override fun onError(error: ErrorResponse){
                onError(error)
            }
        })
    }

    private fun processSection(it: Section): Map<String, Any>? {
        if(it.id != null && it.name != null && it.description != null){
            return mapOf(
                "id" to it.id!!,
                "title" to it.name!!,
                "description" to it.description!!
            )
        }
        return null
    }

    private fun getArticles(section: Long, onSuccess: ListCallback<ArrayList<Map<String,Any>?>>, onError: ErrorCallback){
        helpCenterProvider.getArticles( section, object: ZendeskCallback<List<Article>>(){
            override fun onSuccess(articles: List<Article>){
                val articleList : ArrayList<Map<String, Any>?> = ArrayList<Map<String, Any>?>()
                articles.forEach{
                    articleList.add(processArticle(it))
                }
                onSuccess(articleList)
            }
            override fun onError(error: ErrorResponse){
                onError(error)
            }
        })
    }

    private fun getArticle(articleId: Long, onSuccess: SingleCallback<Map<String,Any>?>, onError: ErrorCallback){
        helpCenterProvider.getArticle( articleId, object: ZendeskCallback<Article>(){
            override fun onSuccess(article: Article){
                onSuccess(processArticle(article))
            }
            override fun onError(error: ErrorResponse){
                onError(error)
            }
        })
    }

    private fun processArticle(it: Article): Map<String, Any>? {
        if(it.id != null && it.updatedAt != null && it.body != null && it.title != null){
            return mapOf(
                "id"  to it.id!!,
                "title"  to it.title!!,
                "content"  to it.body!!,
                "updatedAt"  to it.updatedAt!!,
            )
        }
        return null
    }
}

File: ios/Runner/ZendeskApi.swift

//
//  ZendeskApi.swift
//  Runner
//
//  Created by Ricardo on 13/12/21.
//

import Foundation
import Flutter
import SupportSDK
import ZendeskCoreSDK

class ZendeskApi {

    private var helpcenterProvider : ZDKHelpCenterProvider;
    
    public init() {
        CoreLogger.enabled = true
        CoreLogger.logLevel = .debug
        Zendesk.initialize(
            appId: "12abbc10c46d1731876d529d0b90656cc81d416a9452d0be",
            clientId: "mobile_sdk_client_3beab601d34d787fc05c",
            zendeskUrl: "https://mytikihelp.zendesk.com/"
        )
        let identity = Identity.createAnonymous()
        helpcenterProvider = ZDKHelpCenterProvider()
        Zendesk.instance?.setIdentity(identity)

    }
    
    func getZendeskCategories(result:  @escaping FlutterResult){
        getCategories(
            onSuccess: {categories -> Void in result(categories)},
            onError: {error -> Void in
                result(FlutterError.init(code: "-1", message: error?.localizedDescription ?? "Generic error on line 36 ZendeskApi", details: nil))
            }
       )
    }
    
    func getZendeskSections(call: FlutterMethodCall, result:  @escaping FlutterResult){
        guard let args = call.arguments as? Dictionary<String, Any>,
              let categoryId = args["categoryId"] as? NSNumber else {
                  result(FlutterError.init(code:"400", message:"categoryId argument is missing", details: nil))
                  return
              }
        getSections(
            categoryId: categoryId.stringValue,
            onSuccess: {sections -> Void in result(sections)},
            onError: {error -> Void in
                result(FlutterError.init(code: "-1", message: error?.localizedDescription ?? "Generic error on line 52 ZendeskApi", details: nil))
            }
       )
    }
    
    func getZendeskArticles(call: FlutterMethodCall, result:  @escaping FlutterResult){
        guard let args = call.arguments as? Dictionary<String, Any>,
              let sectionId = args["sectionId"] as? NSNumber else{
                  result(FlutterError.init(code:"400", message:"sectionId argument is missing", details: nil))
                  return
        }
        getArticles(
            sectionId: sectionId.stringValue,
            onSuccess: {articles -> Void in result(articles)},
            onError: {error -> Void in
                result(FlutterError.init(code: "-1", message: error?.localizedDescription ?? "Generic error on line 67 ZendeskApi", details: nil))
            }
       )
    }
    
    func getZendeskArticle(call: FlutterMethodCall, result: @escaping FlutterResult){
        guard let args = call.arguments as? Dictionary<String, Any>,
              let articleId = args["articleId"] as? NSNumber else{
                  result(FlutterError.init(code:"400", message: "articleId argument is missing", details: nil))
            return
        }
        getArticle(
            articleId: articleId.stringValue,
            onSuccess: {article -> Void in result(article)},
            onError: {error -> Void in
                result(FlutterError.init(code: "-1", message: error?.localizedDescription ?? "Generic error on line 82 ZendeskApi", details: nil))
            }
        )
    }
    
    private func getCategories(onSuccess: @escaping  ([[String: Any]]) -> Void,
                               onError: @escaping  (Error?) -> Void){
        helpcenterProvider.getCategoriesWithCallback({ anyOptinoalList, error -> Void in
            guard let anyList = anyOptinoalList as? [Any],
                  let anyObjectList = anyList as? [AnyObject],
                  let categories = anyObjectList as? [ZDKHelpCenterCategory] else{
                      let error = NSError(domain: "com.mytiki.app", code: -1, userInfo: ["en":"Could not convert element to ZDKHelCenterCategory"])
                      onError(error)
                      return
            }
            let result = categories.map { cat -> [String:Any] in
            return [
                    "id" : cat.identifier ?? 0,
                    "title" : cat.name ?? "",
                    "description"  : cat.description
                ]
            }
            onSuccess(result)
        })
    }
    
    private func getSections( categoryId: String,
                              onSuccess: @escaping  ([[String:Any]]) -> Void,
                              onError: @escaping  (Error?) -> Void){
        helpcenterProvider.getSectionsWithCategoryId(categoryId,  withCallback: {sections, error -> Void in
            guard let sections = sections as? [ZDKHelpCenterSection] else {
                onError(error)
                return
            }
            let result = sections.map { (section) -> [String:Any] in
                return [
                    "id" : section.identifier ?? 0,
                    "title" : section.name ?? "",
                    "description"  : section.description
                    ]
            }
            onSuccess(result)
        })
    }
    
    private func getArticles( sectionId: String,
                              onSuccess: @escaping  ([[String:Any]]) -> Void,
                              onError: @escaping  (Error?) -> Void){
        helpcenterProvider.getArticlesWithSectionId(sectionId,  withCallback: {articles, error -> Void in
            guard let articles = articles as? [ZDKHelpCenterArticle] else {
                onError(error)
                return
            }
            let result = articles.map { (article) -> [String:Any] in
                return [
                    "id" : article.identifier ?? 0,
                    "title" : article.title ?? "",
                    "content"  : article.body ?? "",
                    "updatedAt" : self.getStringFromDate(date: article.created_at) ?? ""
                ]
            }
            onSuccess(result)
        })
    }
    
    
    private func getArticle( articleId: String,
                             onSuccess: @escaping  ([String:Any]) -> Void,
                             onError: @escaping  (Error?) -> Void){
        helpcenterProvider.getArticleWithId(articleId, withCallback: {articles, error -> Void in
            guard let articles = articles as? [ZDKHelpCenterArticle] else{
                onError(error)
                return
            }
            let article = articles[0]
            let result : [String:Any] = [
                "id" : article.identifier ?? 0,
                "title" : article.title ?? "",
                "content"  : article.body ?? "",
                "updatedAt" : self.getStringFromDate(date: article.created_at) ?? ""
            ]
            onSuccess(result)
        })
    }
    
    private func getStringFromDate(date: Date) -> String?{
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "EEE MMM dd HH:mm:ss YYYY"
        return dateFormatter.string(from: date)
    }
}
